### DISCUSSION TITLE
Idea to improve the small file reads.

### DISCUSSION THREAD

**Discussion by Jazz9** [minio/minio#17103] [Q&A] [answered]
I've been noticing that our minio cluster does many 32kb reads. After reading the code, I suspect that any file under 1MB is read in 32KB increments. This causes alot of IO and therefore slowness.

I wonder if it would be better to read that whole file under 512KB and move Blocksize small upto 128KB from 32KB for files over 512KB?

	BlockSizeSmall       = 32 * humanize.KiByte // Default r/w block size for smaller objects.
	BlockSizeLarge       = 2 * humanize.MiByte  // Default r/w block size for larger objects.
	BlockSizeReallyLarge = 4 * humanize.MiByte  // Default write block size for objects per shard >= 64MiB

A Secondary option would to make the 32KB Default small block size a tunable? 

This would assist HDD performance and not adversely affect SSD performance IMO.

Thoughts?

**Comment by klauspost** [comment, marked as answer]
It does seem like that could be tweaked a bit.

I will propose:

```Go
	// Select bigger blocks when reading would do more than 2 reads.
	r.SmallFile = st.Size() <= BlockSizeSmall*2
```

This will use a big block for >64KB, which seems reasonable. Syscalls are pretty expensive, so this should minimize it.

We don't add "tweaking" options, unless strictly needed.

**Reply by klauspost** [reply]
PR with change: https://github.com/minio/minio/pull/17107

**Comment by Jazz9** [comment]
Thankyou for the patch - Do you know when it might be released and I can perform benchmarks? Or - is there a guide to compile the current master for testing?

Should the write block size also be increased?

With larger write block size, the fragmentation will be reduced and therefore improve the read speed.

With larger write block size, the seek time of the drives will be reduced. From my analysis of our minio clusters, the drives are seeking the majority of time ( ~92% seek time vs 8% read/write ).

**Reply by klauspost** [reply]
It will be in the next release. There is no fixed schedule for releases, so just monitor them.

0 ->128K uses 32KB, 128K -> 64MB uses 2MB, above 4MB. That is fine. Bigger buffers both use more memory, stresses CPU cache more and increases latency, so bigger buffers are not always better.

**Reply by Jazz9** [reply]
Thank you for the info - I have made a fork to test a few different changes and then I'll report back the effectiveness.

**Comment by Jazz9** [comment]
Should this in xl-storage also be changed to match the new intent of the code?

line 1811
	or := &xioutil.ODirectReader{
		File: file,
		// Select bigger blocks when reading at least 50% of a big block.
		SmallFile: length <= xioutil.BlockSizeLarge/2,
	}

to: 

SmallFile: length <= xioutil.BlockSizeSmall*2

**Reply by klauspost** [reply]
No. This is a streaming read. These can complete at any speed, so if we overallocate buffers here we potentially have a lot of hanging big allocs.

**Reply by Jazz9** [reply]
That makes sense!
